### PR TITLE
Added unicode support

### PR DESCRIPTION
**Pull request by vinsdragonis**
Can parse Unicode literals following the format \u{HHHH...} in a string

### CODE REVIEW

**Comment by vinsdragonis** [comment]
#1982

File: .vscode/extensions.json [hunk 1]
```diff
@@ -1,10 +0,0 @@
-{
```
Comment (jonmeow) [inline .vscode/extensions.json]: It looks like you're deleting this file by accident.

File: common/string_helpers.cpp [hunk 2]
```diff
@@ -81,8 +82,40 @@ auto UnescapeStringLiteral(llvm::StringRef source, const int hashtag_num,
           ret.push_back(16 * *c1 + *c2);
           break;
         }
-        case 'u':
-          CARBON_FATAL() << "\\u is not yet supported in string literals";
+        case 'u': {
+          i++;
+          if (i >= source.size() || source[i] != '{')
+            return std::nullopt;
```
Comment (jonmeow) [inline common/string_helpers.cpp]:
```suggestion
          if (i >= source.size() || source[i] != '{') {
            return std::nullopt;
          }
```

Just a matter of style -- [we always use braces for conditionals](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/cpp_style_guide.md#syntax-and-formatting).

File: common/string_helpers.cpp [hunk 3]
```diff
@@ -81,8 +82,40 @@ auto UnescapeStringLiteral(llvm::StringRef source, const int hashtag_num,
...
+              hex_val = source[i] - 'A' + 10;
+            } else {
+              return std::nullopt;
+            }
+            unicode_int = unicode_int << 4;
+            unicode_int += hex_val;
+            i++;
+          }
+          if (i >= source.size())
+            return std::nullopt;
```
Comment (jonmeow) [inline common/string_helpers.cpp]:
```suggestion
          if (i >= source.size()) {
            return std::nullopt;
          }
```

Style, as above.

File: common/string_helpers_test.cpp [hunk 4]
```diff
@@ -30,6 +30,7 @@ TEST(UnescapeStringLiteral, Valid) {
   EXPECT_THAT(UnescapeStringLiteral("\\x12"), Optional(Eq("\x12")));
   EXPECT_THAT(UnescapeStringLiteral("test", 1), Optional(Eq("test")));
   EXPECT_THAT(UnescapeStringLiteral("test\\#n", 1), Optional(Eq("test\n")));
+  EXPECT_THAT(UnescapeStringLiteral("r\\u{E9}al\\u{2764}\\u{FE0F}\\u{1F50A}!"), Optional(Eq("réal❤️🔊!")));
```
Comment (jonmeow) [inline common/string_helpers_test.cpp]: Please add more invalid test cases for this (see the `Invalid` test below). I think your code handles most of these, but you'll probably find a few issues. It's always good to have tests to check the logic on edge cases. A few I'd highlight are:


```
// Various incomplete cases.
"\\u"
"\\uz"
"\\u{"
"\\u{z"
"\\u{E9"
"\\u{E9z"
// Lowercase hex.
"\\u{e9}"
// Expecting this to be rejected because it's past the maximum codepoint.
// ConvertCodePointToUTF8 should fail.
"\\u{110000}"
// More than 32 bits, but may incorrectly calculate to E9 when stored in a 32-bit unsigned.
"\\u{FF000000E9}"
```

Also, one more valid case that I think can be tested:

```
// Maximum valid codepoint.
"\\u{10FFFF}"
// Nothing wrong with this, just a peculiar case.
"\\u{0000}"
```

File: common/string_helpers.cpp [hunk 5]
```diff
@@ -81,8 +82,40 @@ auto UnescapeStringLiteral(llvm::StringRef source, const int hashtag_num,
...
+          char temp[4];
+          char *result_two = &temp[0];
+          char *result = &temp[0];
+          bool conversion_ok = llvm::ConvertCodePointToUTF8(unicode_int, result_two);
+          if (!conversion_ok) {
```
Comment (jonmeow) [inline common/string_helpers.cpp]:
```suggestion
          if (!llvm::ConvertCodePointToUTF8(unicode_int, result_two)) {
```

There's no need for a separate variable, we can test the function return directly. Minimizing variables helps avoid the question of whether it's used again later.

File: common/string_helpers.cpp [hunk 6]
```diff
@@ -81,8 +82,40 @@ auto UnescapeStringLiteral(llvm::StringRef source, const int hashtag_num,
...
+            return std::nullopt;
+          }
+          while(result < result_two) {
+            ret.push_back(*result);
+            result++;
+          }
```
Comment (jonmeow) [inline common/string_helpers.cpp]:
```suggestion
          ret.append(temp, result_two - temp);
```

Note we can use pointer math and `append` to avoid writing a loop. With this, I think you can also eliminate the `result` pointer.

File: common/string_helpers.cpp [hunk 7]
```diff
@@ -81,8 +82,40 @@ auto UnescapeStringLiteral(llvm::StringRef source, const int hashtag_num,
...
+          i++;
+          while (i < source.size() && source[i] != '}') {
+            char hex_val;
+            if (source[i] >= '0' && source[i] <= '9') {
+              hex_val = source[i] - '0';
+            } else if (source[i] >= 'A' && source[i] <= 'F') {
+              hex_val = source[i] - 'A' + 10;
+            } else {
+              return std::nullopt;
+            }
```
Comment (jonmeow) [inline common/string_helpers.cpp]: There's a `FromHex` helper function that you can probably use here.

File: common/string_helpers.cpp [hunk 8]
```diff
@@ -81,8 +82,40 @@ auto UnescapeStringLiteral(llvm::StringRef source, const int hashtag_num,
           ret.push_back(16 * *c1 + *c2);
           break;
         }
-        case 'u':
-          CARBON_FATAL() << "\\u is not yet supported in string literals";
+        case 'u': {
+          i++;
```
Comment (jonmeow) [inline common/string_helpers.cpp]: Style nit: prefer preincrement (`++i`) unless you specifically need postincrement semantics.
https://google.github.io/styleguide/cppguide.html#Preincrement_and_Predecrement

File: common/string_helpers.cpp [hunk 9]
```diff
@@ -81,8 +82,40 @@ auto UnescapeStringLiteral(llvm::StringRef source, const int hashtag_num,
...
+            }
+            unicode_int = unicode_int << 4;
+            unicode_int += hex_val;
+            i++;
+          }
+          if (i >= source.size())
+            return std::nullopt;
+          char temp[4];
+          char *result_two = &temp[0];
+          char *result = &temp[0];
```
Comment (jonmeow) [inline common/string_helpers.cpp]: It may help readability to clarify what this buffer is going to hold, e.g. calling it `utf8` instead of `temp`. Also, it may help to call `result_two` `utf8_end` to emphasize how it's going to be used.

File: common/string_helpers_test.cpp (hunk 4 above)
Comment (jonmeow) [inline common/string_helpers_test.cpp]: After looking between this and the design, I've filed https://github.com/carbon-language/carbon-lang/pull/2040 to propose a restriction on the number of digits.

Because that proposal isn't yet accepted, I'd suggest adding tests for both `\u{}` and `\u{00000000E9}`. However, whether those are valid (matching the current design) or invalid (matching #2040) I think is up to you -- regardless, the behavior should be tested and consistent.

**Review by jonmeow — COMMENTED** [review]
Thanks for writing this! Looking through the implementation, I think it needs to handle a few more edge cases. I've suggested some tests that you should add to help see what passes and fails, and to get a better idea of what to handle.

I've also made a few style comments, and please don't mind those much, they're just to make sure our code style is consistent.

File: .vscode/extensions.json (hunk 1 above)
Comment (vinsdragonis) [reply .vscode/extensions.json]: I've seen that happen a few times when people use VS Code with a different set of extensions. If it was deleted, it's probably by accident, yeah. Sorry about that.

File: common/string_helpers.cpp (hunk 6 above)
Comment (vinsdragonis) [reply common/string_helpers.cpp]: This change threw an error for me while testing, so I will _temporarily_ revert this back. I'll check on why it didn't work because your suggestion may be valid, but it may need a few more modifications to get it to work.

Is it because you missed the `result` variable? That's what the error message showed, so I am assuming that to be the case. Please correct me if I am wrong on that.

File: common/string_helpers.cpp (hunk 7 above)
Comment (vinsdragonis) [reply common/string_helpers.cpp]: Thanks, I'll check it out and see if I can put it to use here.

File: common/string_helpers.cpp (hunk 8 above)
Comment (vinsdragonis) [reply common/string_helpers.cpp]: It doesn't show any difference at the moment, but I'll change it anyways.

File: common/string_helpers.cpp (hunk 9 above)
Comment (vinsdragonis) [reply common/string_helpers.cpp]: As mentioned previously, I am not using it right now since it threw an unusual error. But I will check back on this.

File: common/string_helpers_test.cpp (hunk 4 above)
Comment (vinsdragonis) [reply common/string_helpers_test.cpp]: Added test cases for the above-mentioned. Do let me know if there needs to be any modification.

Comment (vinsdragonis) [reply common/string_helpers_test.cpp]: Invalid test cases:
- `\u{e9}`
- `\u{E9z}`
- `\uz`
- `\u{`
- `\u`

The above test cases haven't been added to the program yet. Do let me know if I should add them.
I have added a few more requested test cases to the program now. They are valid.

File: .vscode/extensions.json (hunk 1 above)
Comment (jonmeow) [reply .vscode/extensions.json]: Looks like the delete is still here, and now it's adding .vscode/configurationCache.log (that one, we have /.vscode/settings.json in .gitignore)

File: common/string_helpers.cpp (hunk 6 above)
Comment (Pritjam) [reply common/string_helpers.cpp]: @jonmeow , thanks for the suggestion! I was originally trying to do something like this, but I guess I was rusty on pointers, because I couldn't quite implement it.

File: common/string_helpers.cpp (hunk 9 above)
Comment (Pritjam) [reply common/string_helpers.cpp]: @jonmeow, thanks for the suggestion. Actually, I was wondering if you could help me understand something with that part of the code. I didn't originally want to make all of these extra pointers. The plan was to somehow pass a pointer to the temp array straight into ConvertCodePointToUTF8. However, I couldn't get it to work, mainly because I didn't properly understand the type being used, which was a char *&. 

From what I understand, that's a reference to a char pointer, so I thought I could just pass in &temp, since temp is a char * (an array, which is really a char pointer). However, that didn't work, and after trying a few more things I eventually got this, which works, but feels unnecessarily messy. Is there a better way to do this?

File: common/string_helpers.cpp (hunk 7 above)
Comment (jonmeow) [reply common/string_helpers.cpp]: This still needs to be addressed.

File: common/string_helpers.cpp [hunk 10]
```diff
@@ -81,8 +82,42 @@ auto UnescapeStringLiteral(llvm::StringRef source, const int hashtag_num,
...
+          if (!llvm::ConvertCodePointToUTF8(unicode_int, result_two)) {
+            return std::nullopt;
+          }
+          while (result < result_two) {
+            ret.push_back(*result);
+            result++;
+          }
```
Comment (jonmeow) [inline common/string_helpers.cpp]:
```suggestion
          result.append(temp, result_two - temp);
```

I must've forgotten to add this suggestion before. Note it removes the need for what's currently `result` by eliminating the loop.

File: common/string_helpers.cpp (hunk 9 above)
Comment (jonmeow) [reply common/string_helpers.cpp]: @vinsdragonis What was the error?

@Pritjam Sorry, I failed to include my `result.append` suggestion (added below). You can't just pass in `&temp` because ConvertCodePointToUTF8 requires a reference parameter to indicate the end of the UTF8 sequence: [documentation](https://llvm.org/doxygen/namespacellvm.html#ad43161495477e1f7f4ec404e2bb9b075)

File: common/string_helpers_test.cpp (hunk 4 above)
Comment (jonmeow) [reply common/string_helpers_test.cpp]: Yes, please add them. It's important that they're tested and handled correctly.

Comment (jonmeow) [reply common/string_helpers_test.cpp]: It looks like #2040 is being accepted, so please make sure those cases are correctly handled as invalid.

File: explorer/testdata/string/unicode.carbon [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+fn CompareStr(s: String) -> i32 {
+  if (s == "HELLO WORLD!!") {
+    return 0;
+  }
+  return 1;
+}
+
+fn Main() -> i32 {
+  Print("\u{0048}\u{0045}\u{004C}\u{004C}\u{004F} \u{0057}\u{004F}\u{0052}\u{004C}\u{0044}\u{0021}\u{21}\u{1F601}");
+  Print("\u{10FFFF}\u{0000}\u{00000000E9}\u{}");
```
Comment (jonmeow) [inline explorer/testdata/string/unicode.carbon]: Note, printing null characters is going to mess with output. Please don't do this -- unit testing is better for checking things like this.

**Review by jonmeow — CHANGES_REQUESTED** [review]

File: common/string_helpers.cpp (hunk 9 above)
Comment (vinsdragonis) [reply common/string_helpers.cpp]: I managed to get this working, though I am not exactly sure this is how it is meant to be. After eliminating the previously mentioned while loop, I got rid of `char *result = &temp[0];` as it was not used anywhere else. That seemed to fix the problem for me.

File: common/string_helpers.cpp (hunk 6 above)
Comment (vinsdragonis) [reply common/string_helpers.cpp]: I managed to get this working by eliminating `char *result = &temp[0];`, though I am not sure if that's what was supposed to be done.

File: common/string_helpers_test.cpp (hunk 4 above)
Comment (vinsdragonis) [reply common/string_helpers_test.cpp]: Added these tests to another file called fail_unicode.carbon

File: explorer/testdata/string/unicode.carbon (hunk 11 above)
Comment (vinsdragonis) [reply explorer/testdata/string/unicode.carbon]: I have shifted them to a new file, just to check if they are accepted or not.

Comment (Pritjam) [reply explorer/testdata/string/unicode.carbon]: I think a better approach would be to write unit tests. If I understand correctly, this means writing more tests in the format shown in common/string_helpers_test.cpp. I've written a few such tests, and I will add them to your fork in a bit once I finish writing them @vinsdragonis

File: common/string_helpers.cpp (hunk 7 above)
Comment (vinsdragonis) [reply common/string_helpers.cpp]: Doing this now.

File: explorer/testdata/string/unicode.carbon (hunk 11 above)
Comment (Pritjam) [reply explorer/testdata/string/unicode.carbon]: I've finished my tests. @vinsdragonis , if you could add me as a collaborator to your fork, I can push them.

Comment (vinsdragonis) [reply explorer/testdata/string/unicode.carbon]: Sent invite

File: .vscode/extensions.json (hunk 1 above)
Comment (Pritjam) [reply .vscode/extensions.json]: So, just to confirm, we should add back the extensions.json file, but remove the configurationCache?

Edit: I've added back the extensions.json file.

File: common/string_helpers_test.cpp (hunk 4 above)
Comment (Pritjam) [reply common/string_helpers_test.cpp]: The tests in common/string_helpers_test.cpp, in the block TEST(UnescapeStringLiteral, InvalidUnicodes) should address these as unit tests.

File: explorer/testdata/string/unicode.carbon (hunk 11 above)
Comment (Pritjam) [reply explorer/testdata/string/unicode.carbon]: Looks like we're still having some errors, and I think those are because of the test fail_unicode.carbon. I think we can remove this test, since all of the features it tests are already in the unit tests I added.

Comment (vinsdragonis) [reply explorer/testdata/string/unicode.carbon]: Figures why I had a problem with the passing of the checks. I'll remove it.

Comment (vinsdragonis) [reply explorer/testdata/string/unicode.carbon]: @Pritjam you were right. The checks didn't pass only because of `fail_unicode.carbon`.

File: common/string_helpers.cpp (hunk 7 above)
Comment (vinsdragonis) [reply common/string_helpers.cpp]: This has been resolved by @Pritjam

File: .vscode/extensions.json (hunk 1 above)
Comment (jonmeow) [reply .vscode/extensions.json]: You're still modifying this file. Please restore the original version.

File: common/string_helpers_test.cpp (hunk 4 above)
Comment (jonmeow) [reply common/string_helpers_test.cpp]: > ```
> // Various incomplete cases.
> "\\u"
> "\\uz"
> "\\u{"
> "\\u{z"
> "\\u{E9"
> "\\u{E9z"
> // Lowercase hex.
> "\\u{e9}"
> // Expecting this to be rejected because it's past the maximum codepoint.
> // ConvertCodePointToUTF8 should fail.
> "\\u{110000}"
> // More than 32 bits, but may incorrectly calculate to E9 when stored in a 32-bit unsigned.
> "\\u{FF000000E9}"
> ```
> 
> Also, one more valid case that I think can be tested:
> 
> ```
> // Maximum valid codepoint.
> "\\u{10FFFF}"
> // Nothing wrong with this, just a peculiar case.
> "\\u{0000}"
> ```

You seem to have missed several of these test cases, and importantly tests like `\\u{z` which I think the code still doesn't handle correctly. It's not ensuring there's a closing brace.

Test this _entire list_ of test cases. Testing edge cases is very important. Part of the reason I'm providing a list of test cases is to help you see and fix the bugs in your code; it's very important to add them when I'm asking you to.

Additionally, given #2040 and what you're testing, I'll note an additional test case that you need:

`"\\u000000E9"`

This should be valid, while the current code renders it invalid.

File: .vscode/extensions.json (hunk 1 above)
Comment (vinsdragonis) [reply .vscode/extensions.json]: Copied back the original contents to it, the way it was.

Comment (Pritjam) [reply .vscode/extensions.json]: I believe we have to actually revert it with a git checkout, so no changes happen. Is that right @jonmeow ?

**Review by jonmeow — CHANGES_REQUESTED** [review]
Let me know when requested tests have all been added and are passing.

File: common/string_helpers_test.cpp (hunk 4 above)
Comment (Pritjam) [reply common/string_helpers_test.cpp]: Sorry about that, I must have missed those. I've added tests for all of them, except for two:

-For the zero codepoint `\u{0000}`, I'm having trouble writing a test case, since any attempt to put `\0` or any other form of "zero" in a C++ string just results in it terminating prematurely, because C++ uses 0 for string terminators. How should I write a test case for this one?

-For the `"\\u000000E9"`, I didn't add one exactly like that, since by my understanding of Carbon's Unicode semantics, there should be curly braces. I did add such a case with curly braces to the code though, in the "réal" test case, and I changed the logic in the unicode parsing code to check for a limit of 8 hex chars. Please let me know if those work ok.

File: .vscode/extensions.json (hunk 1 above)
Comment (vinsdragonis) [reply .vscode/extensions.json]: @Pritjam, is this file restored to its original state?

Comment (vinsdragonis) [reply .vscode/extensions.json]: @jonmeow, please let us know if by any chance this one is being deleted, as that doesn't appear to be the case on our end.

**Comment by vinsdragonis** [comment]
@jonmeow, other than a confirmation from your side regarding vscode extensions, I think we have covered everything else as requested. Please let us know if we missed something, or if something still needs to be addressed.

**Review by jonmeow — CHANGES_REQUESTED** [review]
It looks like some formatting issues have gotten in, can you look over the errors from pre-commit (https://github.com/carbon-language/carbon-lang/runs/7923463205?check_suite_focus=true) and fix those?

Once that's done, I'm fine approving, thanks!

**Comment by jonmeow** [comment]
Note, if you set up pre-commit per https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/contribution_tools.md#pre-commit, it should fix these automatically. It just can't safely do that on github.

**Review by jonmeow — APPROVED** [review]

**Comment by jonmeow** [comment]
Thanks! I've merged the PR in.
